src/presentation/api/controller/services.go

```diff
@@ -0,0 +1,212 @@
+package apiController
+
+import (
+	"log/slog"
+
+	"github.com/labstack/echo/v4"
+	voHelper "github.com/speedianet/os/src/domain/valueObject/helper"
+	internalDbInfra "github.com/speedianet/os/src/infra/internalDatabase"
+	apiHelper "github.com/speedianet/os/src/presentation/api/helper"
+	"github.com/speedianet/os/src/presentation/service"
+)
+
+type ServicesController struct {
+	servicesService *service.ServicesService
+}
+
+func NewServicesController(
+	persistentDbService *internalDbInfra.PersistentDatabaseService,
+) *ServicesController {
+	return &ServicesController{
+		servicesService: service.NewServicesService(persistentDbService),
+	}
+}
+
+// ReadServices	 godoc
+// @Summary      ReadServices
+// @Description  List installed services and their status.
+// @Tags         services
+// @Security     Bearer
+// @Accept       json
+// @Produce      json
+// @Success      200 {array} dto.InstalledServiceWithMetrics
+// @Router       /v1/services/ [get]
+func (controller *ServicesController) Read(c echo.Context) error {
+	return apiHelper.ServiceResponseWrapper(c, controller.servicesService.Read())
+}
+
+// ReadInstallableServices	 godoc
+// @Summary      ReadInstallableServices
+// @Description  List installable services.
+// @Tags         services
+// @Security     Bearer
+// @Accept       json
+// @Produce      json
+// @Success      200 {array} entity.InstallableService
+// @Router       /v1/services/installables/ [get]
+func (controller *ServicesController) ReadInstallables(c echo.Context) error {
+	return apiHelper.ServiceResponseWrapper(
+		c, controller.servicesService.ReadInstallables(),
+	)
+}
+
+func parseRawPortBindings(bindings []interface{}) []string {
+	rawPortBindings := []string{}
+	for _, rawPortBinding := range bindings {
+		rawPortBindingMap, assertOk := rawPortBinding.(map[string]interface{})
+		if !assertOk {
+			slog.Debug(
+				"InvalidPortBindingStructure", slog.Any("portBinding", rawPortBinding),
+			)
+			continue
+		}
+
+		rawPortStr, err := voHelper.InterfaceToString(rawPortBindingMap["port"])
+		if err != nil {
+			slog.Debug(err.Error(), slog.String("port", rawPortStr))
+			continue
+		}
+
+		rawProtocolStr, err := voHelper.InterfaceToString(rawPortBindingMap["protocol"])
```

Comment:
Protocol may or may not be sent, no? Thus you can't assume it's there.

Reply:
Yeah, you're right.
I added validation before creating the **VO** and decided to pass the default value as empty. The **PortBinding VO** itself will recognize the absence of a protocol and add **TCP** by default.